### ISSUE TITLE
addItemToOrder race condition

### ISSUE THREAD

**Issue by ondratuma** [vendure-ecommerce/vendure#2187]
**Describe the bug**
When there are multiple concurrent calls to addItemToOrder, it results in undefined behavior, where some items are not added.

**To Reproduce**
We are adding order lines like this, due to the custom fields
```js
const responses = await Promise.all(
      tipValues.map(async (tipValue: any) => {
        return api.order.mutations.addItemToOrder({
          productVariantId,
          quantity: 1,
          customFields: { tipValueAbsolute: tipValue },
        });
      })
    );}
```

**Expected behavior**
All items should be in the order if all calls succeeded.

**Environment (please complete the following information):**
 - @vendure/core version: 1.9.2
 - Nodejs version: 16.x
 - Database: postgres

**Additional context**

**Comment by ondratuma** [comment]
If concurrent calls do cause such behavior, it is perhaps required to aquire pesimistic lock on the order entity ?

**Comment by michaelbromley** [comment]
Updates from Discord:

> We've now added 300ms delay after each mutation and its working as intended

> These are the mutations we are calling upon checkout. We are currently sending them separately (as I said, awaiting response + 300ms delay). We were also thinking about sending them as one query.

- setOrderShippingAddress
- setCustomerForOrder
- custom resolver that sets the shippingMethod
 Ultimately just calling: await this.orderService.setShippingMethod(ctx, session.activeOrderId, [shippingMethod.id]);
- transition to arringingPayment
- addPaymentToOrder

**Comment by ondratuma** [comment]
@michaelbromley 

This issue would arise, even if you lets say, wanted to have multiple users editing orders at the same time (wholesale).

The way I see this issue usually handled, is using some shared lock, if we do not want to handle this on a database level.
It could be configurable strategy, where I think the default, for app running on one server, in memory would be enough, and the implementation simple.

For advanced use-cases, redis could be used.

Consumer of the api, especially if the method is called addItemToOrder, should not care about the race-conditions at all from my point of view.

**Comment by ondratuma** [comment]
It would generally wait for a certain period, and then fail.
But any approach (even failing instantly), is better than race condition.

Some, partially feasible approach could be similar to https://github.com/felanios/murlock

The problem I can see, is with the usage of implicit transaction with the Transaction decorator.
As the locks should be acquire before starting the transactions, otherwise the lock would serve no purpose with the default isolation levels (mysql/postgres).

So the approach I am used to is to 
- fetch the resource id before the transaction. (orderId based on user session)
- acquire the lock on the resource (order in most cases)
- start transaction
- perform action
- commit the transaction

The approach I am used to is to add another layer between controllers and services, where transactions would be started. Basically duplication of controller methods. 

_Another approach could be adding optimistic lock (@VersionColumn) on the order table, that should resolve the data consistency for most of the cases that might affect the customer directly through shop-api (I assume). 
It would not resolve the cases that do not modify the order directly. 
With automatic transaction retries, the experience for most actions would be the same as waiting for a lock aquisition._

_Currently, I do not have deep enough knowledge of all the possible data races that can happen, to properly evaluate the locking strategy. 
With the shared locks being possibly configurable as opt-in/opt-out, I do not see a problem with introducing the funccionality, as it would not even require a migration._

I would be happy to collaborate.

**Comment by williamrijksen** [comment]
Hi all,

I ran into the same issue with concurrency when adding items to an order. I implemented a custom mutation that uses a pessimistic write lock on the order before adding items.

The general approach:
- Fetch the active order via `activeOrderService.getActiveOrder`.
- Acquire a pessimistic write lock on the order row in the database.
- Call `orderService.addItemToOrder` to safely add the item.

This ensures that concurrent requests modifying the same order don’t conflict.

I can share the full mutation code if needed. Curious if this aligns with Vendure’s recommended way to handle this.
